workerbee/postgresql.py

```diff
@@ -1,284 +1,394 @@
+            self.logger.error(msg)
+            raise ValueError(msg)
+        else:
             try:
-                job_function(a_row.job_id, a_row.data if a_row.data is not None else {})
-            except JobFailed as e:
-                sys.stdout.write('FAILED.\n')
-                sys.stdout.flush()
-            else:
-                set_job_as_complete(db_handle, experiment_id, a_row.job_id)
-                if verbose:
-                    sys.stdout.write('...done.\n')
-                    sys.stdout.flush()
-            if verbose:
-                i += 1
-                if i % 10 == 0:
-                    n_completed_rows = get_completed_job_count(db_handle,
-                                                               experiment_id)
-                    print('{:.2f}% ({}/{}) Completed'.format(
-                        (n_completed_rows / n_inputs) * 100,
-                        n_completed_rows, n_inputs))
-    except JobsExhaustedError:
-        sys.stdout.write('\rAll jobs are exhausted, terminating.')
-        sys.stdout.flush()
+                create_table(self.db_handle, self.jobset_id)
+                self.logger.info("Created table for jobset '{}'".format(
+                    self.jobset_id))
+            except psycopg2.ProgrammingError as e:
+                if not ignore_existing_jobset:
+                    # Re-raise - otherwise swallow
+                    raise e
+                else:
+                    self.logger.warn("Table already exists for jobset '{}' - "
+                                     "but ignoring".format(self.jobset_id))
+
+    def does_jobset_exist(self):
+        return table_exists(self.db_handle, self.jobset_id)
+
+    def add_job(self, input_data, cursor=None):
+        query_str = INSERT_JOB_QUERY.format(tbl_name=self.jobset_id)
+        params = {'parameters': {'input_data': postgres_jsonify(input_data)}}
+        if cursor is None:
+            self.db_handle.run(query_str, **params)
+        else:
+            cursor.run(query_str, **params)
+
+    def add_jobs(self, input_datas, cursor=None):
+        if cursor is None:
+            with self.db_handle.get_cursor() as cursor:
+                for input_data in input_datas:
+                    self.add_job(input_data, cursor=cursor)
+        else:
+            for input_data in input_datas:
+                self.add_job(input_data, cursor=cursor)
+        self.logger.info('Submitted {} jobs'.format(len(input_datas)))
+
+    def set_job_as_complete(self, job_id, duration, output_data=None):
+        if output_data is not None:
+            output_data = postgres_jsonify(output_data)
+        params = {
+            'id': job_id,
+            'job_duration': duration,
+            'output_data': output_data
+        }
+        self.db_handle.run(SET_ROW_COMPLETED_BY_ID_QUERY.format(
+            tbl_name=self.jobset_id), parameters=params)
+
+    def get_uncompleted_unclaimed_job(self):
```

Comment:
Maybe these methods should be underscored?

- gives us flexibility to change internal API
- simplifies tab completion suggestions down to public API only